File: COMSOL/cw391_liion_3d_p2d_utilization/InspectMaterials.java
import com.comsol.model.*;
import com.comsol.model.util.*;

public class InspectMaterials {
  private static final String WORKDIR =
      "D:/Users/hez/Desktop/hithium/COMSOL/cw391_liion_3d_p2d_utilization";

  private static void printProp(Model model, String mat, String pg, String prop) {
    try {
      String value = model.component("comp1").material(mat).propertyGroup(pg).getString(prop);
      System.out.println(mat + "." + pg + "." + prop + " = " + value);
    } catch (Exception ex) {
      // Continue.
    }
    try {
      String[] values = model.component("comp1").material(mat).propertyGroup(pg).getStringArray(prop);
      System.out.print(mat + "." + pg + "." + prop + "[] =");
      for (String value : values) {
        System.out.print(" [" + value + "]");
      }
      System.out.println();
    } catch (Exception ex) {
      // Continue.
    }
  }

  public static void main(String[] args) throws Exception {
    ModelUtil.showProgress(false);
    Model model = ModelUtil.load("Model", WORKDIR + "/pouch_cell_utilization_base.mph");
    for (String mat : model.component("comp1").material().tags()) {
      System.out.println("## " + mat + " label=" + model.component("comp1").material(mat).label());
      try {
        int[] domains = model.component("comp1").material(mat).selection().entities(3);
        System.out.print(mat + ".domains =");
        for (int domain : domains) {
          System.out.print(" " + domain);
        }
        System.out.println();
      } catch (Exception ex) {
        System.out.println(mat + ".domains = <unavailable> " + ex.getMessage());
      }
      for (String pg : model.component("comp1").material(mat).propertyGroup().tags()) {
        System.out.println("propertyGroup " + pg);
        try {
          for (String prop : model.component("comp1").material(mat).propertyGroup(pg).properties()) {
            printProp(model, mat, pg, prop);
          }
        } catch (Exception ex) {
          System.out.println(pg + ".properties = <unavailable>");
        }
      }
    }
    ModelUtil.remove(model.tag());
  }
}

File: work/AI_virtual_cell/202609_何争_扣电与软包力电耦合V1/99_内部文件/历史运行归档/COMSOL_runs/20260811_方案1_phis0ec_diagnosis/inspect_materials.py
import json
from pathlib import Path
from com.comsol.model.util import ModelUtil

source = r"C:\HithiumSSD\hithium\work\MIC_LDS\202608_何争_不同集流体导出方式对电芯性能的影响\方案1.mph"
output = Path(r"C:\HithiumSSD\hithium\COMSOL\runs\20260811_方案1_phis0ec_diagnosis\materials_result.json")
tag = "plan1_materials"

def read_property(node, name):
    for method in ("getString", "getStringArray", "getStringMatrix"):
        try:
            value = getattr(node, method)(name)
            if value is not None:
                try:
                    return [list(row) if hasattr(row, "__iter__") and not isinstance(row, str) else row for row in value]
                except TypeError:
                    return str(value)
        except Exception:
            pass
    return None

try:
    ModelUtil.remove(tag)
except Exception:
    pass

model = ModelUtil.load(tag, source)
try:
    comp = model.component("comp1")
    report = {}
    for material_tag in [str(v) for v in comp.material().tags()]:
        material = comp.material(material_tag)
        item = {
            "label": str(material.label()),
            "selection": [int(v) for v in material.selection().entities()],
            "property_groups": {},
        }
        for group_tag in [str(v) for v in material.propertyGroup().tags()]:
            group = material.propertyGroup(group_tag)
            props = {}
            for name in [str(v) for v in group.properties()]:
                value = read_property(group, name)
                if value not in (None, "", []):
                    props[name] = value
            item["property_groups"][group_tag] = {
                "label": str(group.label()),
                "properties": props,
            }
        report[material_tag] = item
    output.write_text(json.dumps(report, ensure_ascii=False, indent=2), encoding="utf-8")
    print(str(output))
finally:
    ModelUtil.remove(tag)
